compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -187,6 +188,10 @@ struct EvalCtxt<'a, 'tcx> {
     var_values: CanonicalVarValues<'tcx>,
 
     search_graph: &'a mut search_graph::SearchGraph<'tcx>,
+
+    /// This field is used by a debug assertion in [`EvalCtxt::evaluate_goal`],
+    /// see the comment in that method for more details.
+    in_projection_eq_hack: bool,
```

Comment:
Thiis hack would be also useful for other `let (_, certainty) = evaluate_goal()` calls in candidate assembly, maybe generalize the name and put it in a helper like `evaluate_goal_expect_stable`?

Reply:
why would it be useful there? All other `evaluate_goal` shouldn't result in cycles when rerunning with substitutions applied (that's at least what I believe and why I've added the debug assert in the first place)